contracts/KyberOasisReserve.sol

```diff
@@ -0,0 +1,258 @@
+        address kyberNetwork,
+        OasisDirectInterface oasisDirect,
+        OtcInterface otc,
+        ERC20 wethToken,
+        ERC20 tradeToken,
+        uint feeBps
+    );
+
+    function setReserveParams(
+        address _kyberNetwork,
+        OasisDirectInterface _oasisDirect,
+        OtcInterface _otc,
+        ERC20 _wethToken,
+        ERC20 _tradeToken,
+        uint _feeBps
+    )
+        public
+        onlyAdmin
+    {
+        require(_kyberNetwork != address(0));
+        require(_oasisDirect != address(0));
+        require(_otc != address(0));
+        require(_wethToken != address(0));
+        require(_tradeToken != address(0));
+        require(_feeBps < 10000);
+
+        kyberNetwork = _kyberNetwork;
+        oasisDirect = _oasisDirect;
+        otc = _otc;
+        wethToken = _wethToken;
+        tradeToken = _tradeToken;
+        feeBps = _feeBps;
+
+        ReserveParamsSet(kyberNetwork, oasisDirect, otc, wethToken, tradeToken, feeBps);
+    }
+
+    function getDestQty(ERC20 src, ERC20 dest, uint srcQty, uint rate) public view returns(uint) {
+        uint dstDecimals = getDecimals(dest);
+        uint srcDecimals = getDecimals(src);
+
+        return calcDstQty(srcQty, srcDecimals, dstDecimals, rate);
+    }
+
+    function valueAfterReducingFee(uint val) public view returns(uint) {
+        require(val <= MAX_QTY);
+        return ((10000 - feeBps) * val) / 10000;
+    }
+
+    function getConversionRate(ERC20 src, ERC20 dest, uint srcQty, uint blockNumber) public view returns(uint) {
+        uint  rate;
+        uint  destQty;
+        ERC20 wrappedSrc;
+        ERC20 wrappedDest;
+        uint  actualSrcQty;
+        uint  actualDestQty;
+        bool  sellEth;
+
+        blockNumber;
+
+        if (!tradeEnabled) return 0;
+        if ((tradeToken != src) && (tradeToken != dest)) return 0;
+
+        sellEth = (src == ETH_TOKEN_ADDRESS);
+
+        if (sellEth) {
+            wrappedSrc = wethToken;
+            wrappedDest = dest;
+            actualSrcQty = valueAfterReducingFee(srcQty);
+        } else if (dest == ETH_TOKEN_ADDRESS) {
+            wrappedSrc = src;
+            wrappedDest = wethToken;
+            actualSrcQty = srcQty;
+        } else {
+            return 0;
+        }
+
+        destQty = otc.getBuyAmount(wrappedDest, wrappedSrc, actualSrcQty);
+
+        if (sellEth) {
+            actualDestQty = destQty;
+        } else {
+            actualDestQty = valueAfterReducingFee(destQty);
+        }
+
+        require(actualDestQty < MAX_QTY);
+        rate = actualDestQty * PRECISION / srcQty;
+
+        return rate;
+    }
+
+    function doTrade(
+        ERC20 srcToken,
+        uint srcAmount,
+        ERC20 destToken,
+        address destAddress,
+        uint conversionRate,
+        bool validate
+    )
+        internal
+        returns(bool)
+    {
+
+        uint actualDestAmount;
+
+        require((ETH_TOKEN_ADDRESS == srcToken) || (ETH_TOKEN_ADDRESS == destToken));
```

Comment:
I think its nicer if requires are on start of function.
more readable for me. not sure if we should decide on convention.
also notice. the sooner the require is, it saves user gas if it fails...

Reply:
done.